### PR TITLE
Python: model fabric

### PR DESCRIPTION
**Pull request by RasmusWL**
I intentionally left out the part of modeling [`fabric.api.execute`](https://docs.fabfile.org/en/1.14/api/core/tasks.html#fabric.tasks.execute) (from version 1.x), since I didn't see an obvious way to integrate this as an additional taint step in our current setup.

@yoff proposed that we might be able to add this in the same way as we do for special methods (such as `__add__`). However, since these are [currently](https://github.com/github/codeql/blob/4ce41854a4de4835a6668909521b940633fb41b2/python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll#L564) as a part of the algebraic datatype `TDataFlowCall` in `DataFlowPrivate`, that doesn't really make it easy to make such a change entirely in the code modeling fabric (that is `Fabric.qll`).

[The current implementation](https://github.com/github/codeql/blob/f07a7bf8cff4152845a013fa44001247e796b3a0/python/ql/src/semmle/python/security/injection/Command.qll#L235-L271) is able to do this as a simple additional taint step by inspecting the first argument of `fabric.api.execute` and the callable it points to -- and it automatically handles the distinction between passing in functions and bound-methods.

I'm hoping we can enable such easy extension as well, but I currently don't see a clear path forwards, so will leave that for future work.

### Modeling subclasses of `fabric.group.Group`

The most interesting part of this PR is the part where we want to model subclasses of `fabric.group.Group`, and not the class itself. This required some thought for how to do so _(you probably need to have read through the changes in https://github.com/github/codeql/commit/98691fe8ec183f668a4f738e6ec1ec91bc8a6b2b before you're able to understand what I'm talking about here)_

After initially using the following code in `module Group`
```ql
/** A reference to a subclass of `fabric.group.Group` */
abstract class SubclassRef extends DataFlow::Node { }

private class SubclassInstantiation extends SubclassInstanceSource, DataFlow::CfgNode {
  override CallNode node;

  SubclassInstantiation() { node.getFunction() = any(SubclassRef ref).asCfgNode() }
}
```

with this code in `module SerialGroup` and `module ThreadingGroup`:

```ql
class ClassRef extends DataFlow::Node, fabric::group::Group::SubclassRef {
  ClassRef() { this = classRef(DataFlow::TypeTracker::end()) }
}
```

I wasn't too much of fan of that approach. Since we probably need the `SubclassInstanceSource` anyway, and don't really have a specific use for `SubclassRef`, I just went with concrete (QL) subclasses of `SubclassInstanceSource` in each of the modules for the Python subclasses.

I really don't know what the best approach is, so I'm very open to suggestions. I think we'll really have to flesh this out for handling Django responses, since we're interested in the fact that some subclasses provide default values for the content-type, and keeping track of that is important for XSS (since there is no XSS if response is `text/plain`)

### CODE REVIEW

File: python/ql/src/experimental/semmle/python/frameworks/Fabric.qll [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+         *
+         * See https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.Group.run
+         */
+        DataFlow::Node subclassInstanceRunMethod() {
+          result = subclassInstanceRunMethod(DataFlow::TypeTracker::end())
+        }
+      }
+
+      /**
+       * A call to `run`, `sudo` on an instance of a subclass of `fabric.group.Group`.
```
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Fabric.qll:505]: How is `sudo` tracked here?

File: python/ql/src/experimental/semmle/python/frameworks/Fabric.qll [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+          result = subclassInstance(DataFlow::TypeTracker::end())
+        }
+
+        /**
+         * Gets a reference to the `run` method on an instance of a subclass of `fabric.group.Group`.
+         *
+         * See https://docs.fabfile.org/en/2.5/api/group.html#fabric.group.Group.run
+         */
+        private DataFlow::Node subclassInstanceRunMethod(DataFlow::TypeTracker t) {
+          t.startInAttr("run") and
```
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Fabric.qll:488]: Should this include "sudo"?

**Review by yoff — CHANGES_REQUESTED** [review]
Generally looks good, just one minor question.
Also, I was momentarily confused by `GroupSubclassInstanceSource` not being connected to `InstanceSource` (because the latter is for `Connection` rather than `Group`).

File: python/ql/src/experimental/semmle/python/frameworks/Fabric.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/src/experimental/semmle/python/frameworks/Fabric.qll:505]: too much copy paste :neutral_face: there is no `sudo` method on a Group, so I'll just remove that bit :+1:

**Comment by RasmusWL** [comment]
> Also, I was momentarily confused by `GroupSubclassInstanceSource` not being connected to `InstanceSource` (because the latter is for `Connection` rather than `Group`).

Aha. In my mind, we're at a pretty OK point in regards to modeling classes in an extensible way with `ClassName::classRef()`, `ClassName::InstanceSource` (abstract class of data-flow nodes), and `ClassName::instance()` -- that does mean that predicate names have fairly generic names, and that all that separates them is the module they are defined in. If you think that's not too good of a solution, please do speak up now :blush:

**Review by yoff — APPROVED** [review]
Looks good now.
Just to comment on the execute function: The special methods framework would need a slight generalisation, I think, in that it currently always invoke a class method. It might be worth it, if we foresee having to model many library functions that result in some function being called...

**Pull request merged**